tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/MetricsHelper.java

```diff
@@ -212,7 +229,431 @@
+                      dict.put(s, size);
+                    }
+                    Op assertion =
+                        tf.withSubScope("assertShapes-2")
+                            .assertThat(
+                                tf.math.equal(
+                                    tf.shape.size(
+                                        symbol.getOperand(),
+                                        tf.constant(ll.getAndIncrement()),
+                                        TInt64.class),
+                                    tf.constant(size)),
+                                Collections.singletonList(tf.constant(message)));
+                    updateOperations.add(assertion);
+                  });
+        });
+
+    return updateOperations;
+  }
+
+  /**
+   * Returns an op to update the given confusion matrix variables.
+   *
+   * <p>For every pair of values in <code>labels</code> and <code>predictions</code>:
+   *
+   * <pre>
+   * TRUE_POSITIVES:  <code>labels</code> == true and <code>predictions</code> &gt; thresholds
+   * FALSE_POSITIVES: <code>labels</code> == true and <code>predictions</code> &lt;= thresholds
+   * TRUE_NEGATIVES:  <code>labels</code> == false and <code>predictions</code> &lt;= thresholds
+   * FALSE_NEGATIVE:  <code>labels</code> == false and <code>predictions</code> &gt; thresholds
+   * </pre>
+   *
+   * <p>The results will be weighted and added together. When multiple thresholds are provided, we
+   * will repeat the same for every threshold.
+   *
+   * <p>For estimation of these metrics over a stream of data, the function creates an `update_op`
+   * operation that updates the given variables.
+   *
+   * <p>If <code>sampleWeight</code> is <code>null</code>, weights default to 1. Use weights of 0 to
+   * mask values.
+   *
+   * @param tf the TensorFlow Ops
+   * @param variablesToUpdate Map with {@link ConfusionMatrixEnum} values as valid keys and
+   *     corresponding variables to update as values. If <code>multiLabel</code> is false then all
+   *     shapes are (T), where T is the number of thresholds. If <code>multiLabel</code> is true
+   *     then all shapes are (T, C0), where C0 is the number of classes.
+   * @param varInitializers Map with {@link ConfusionMatrixEnum} values as valid keys and
+   *     corresponding initializer Operands to initializer the corresponding variables from
+   *     variablesToUpdate.
+   * @param labels the labels, will be cast to {@link TBool} shape (N, Cx, L1?) where N is the
+   *     number of examples, Cx is zero or more class dimensions, and L1 is a potential extra
+   *     dimension of size 1 that would be squeezed. If <code>multiLabel</code> or if <code>
+   *     labelWeights</code> <code>!= null</code>, then Cx must be a single dimension.
+   * @param predictions the predictions shape (N, Cx, P1?).
+   * @param thresholds thresholds in `the range <code>[0, 1]</code>, or {@link #NEG_INF} (used when
```

Comment:
Should we restrict `thresholds` to rank 1? I'm pretty sure this code (and also the Python implementation) breaks if it has a higher rank.

Reply:
The code has the `oneThresh` TBool Operand to handle this,
though the Python code has the comment: `if we will treat thresholds as one-dimensional (always true as of this writing)`.  Do we want to force it to one, or let the code handle it? It looks like they wrote this logic into Python for some future reason.

Reply:
That comment ("if we will...") is actually one I inserted into the Java code in my PR.

I think the Python code related to `oneThresh` is simply buggy:
* `oneThresh` is false in exactly one case: multi-label and rank of thresholds != 1.
* The only use of `oneThresh` is to force `threshLabelTile` to `1` in that case. (Otherwise it is `numLabels`.)
* The only use of `threshLabelTile` is for `threshTiles`. 
* In the case where `oneThresh` is false:
  * `threshTiles` is [1, N, 1], where `N` is the number of examples. (Otherwise, it is [1, N, numLabels].)
  * `thresholdsReshaped` then has shape (T, 1, T*), where T is the size of the 0th dimension of thresholds and T* is the product of the remaining dimensions of thresholds.
  * `threshTiled` then has shape (T, N, T*).
  * But we end up computing `Operand<TBool> predIsPos = tf.math.greater(predsTiled, threshTiled);`
  * In the multi-label case, `predsTiled` has shape (T, N, numLabels) , so that isn't going to work.